README.md
# NAME

Graphics::Skullplot - Plot the result of an SQL select (e.g. from an emacs shell window)

# VERSION

Version 0.01

# SYNOPSIS

    # To use this from emacs, see scripts/skullplot.el.
    # That elisp code accesses the perl script: scripts/skullplot.pl

    # the code used by skullplot.pl
    my $plot_hints = { indie_count           => $indie_count,
                       dependent_requested   => $dependent_requested,
                       independent_requested => $independent_requested,
                     };
    my %gsp_args = 
      ( input_file   => $dbox_file,
        plot_hints   => $plot_hints, );
    $gsp_args{ working_area } = $working_area if $working_area;
    $gsp_args{ image_viewer } = $image_viewer if $image_viewer;
    my $gsp = Graphics::Skullplot->new( %gsp_args );

    $gsp->show_plot_and_exit();  # does an exec 

# DESCRIPTION

Graphics::Skullplot is a module that works with the result from a database 
select in the common tabular text "data box" format. It has routines 
to generate and display plots of the data in png format.

Internally it uses the [Table::BoxFormat](https://metacpan.org/pod/Table::BoxFormat) module to parse the text table,
and the [Graphics::Skullplot::ClassifyColumns](https://metacpan.org/pod/Graphics::Skullplot::ClassifyColumns) module to determine the types of the columns.

The default image viewer is the ImageMagick "display" command.

The immediate use for this code is to act as the back-end for the included 
Emacs package scripts/skullplot.el, so that database select results 
generated in an emacs shell window can be immediately plotted.  

This elisp code calls scripts/skullplot.pl, which might be used in
other contexts.

# METHODS

- new

    Creates a new Graphics::Skullplot object.
    Object attributes:

    - working\_area

        Scratch location where intermediate files are created.
        Defaults to "/tmp".

    - image\_viewer

        Defaults to 'display', the ImageMagick viewer
        (a dependency on Image::Magick ensures it's available)

- builder methods (largely for internal use)

    builder\_image\_viewer Currently just returns a hardcoded selection
    (the ImageMagick "display" program).

- generate\_output\_filenames

    Example usage: 

        # relies on object settings: "input_file" and "working area"
        my $fn = 
          generate_filenames();
        my $basename = $fn->{ base };
        # full paths to file in $working_area
        my $tsv_file  = $fn->{ tsv };  
        my $png_file  = $fn->{ png };  

- plot\_tsv\_to\_png

    Generate the r-code to plot the tsv file data as the png file.
    Takes one argument, a hash of "field metadata".  

    The file names (tsv, png, plus internal formats) come from the
    "naming" object field.

    Example usages:  

        $self->plot_tsv_to_png( $plot_cols ); 

- generate\_png\_file

    Example usage:

        $self->generate_png_file( $pc, $fn );

    Runs the given plot code (first argument) using the file-name metadata
    (second argument, defaults to object's [naming](https://metacpan.org/pod/naming)), saving the 
    plot as a png file ($fn->{png}).

    This generates a file of R code to run with an Rscript call.
    In debug mode, this generates a standalone unix script. ($DEBUG).

- display\_png\_and\_exit

    Open the given png file in an image viewer
    Defaults to "png" field in object's "naming".

    This internally does an exec: it should be
    the last thing called.

    The image viewer can be set as the second, optional field.
    The default image viewer is ImageMagick's "display".

    Example usage:

        my $naming = $self->naming;
        my $png_file = $naming->{ png };
        $self->display_png_and_exit( $png_file );

- show\_plot\_and\_exit

    The method called by the skullplot.pl script to actually
    plot the data from a "data box format" file, using the 
    plot\_hints.

    It's expected that the dbox file ([input\_file](https://metacpan.org/pod/input_file)) and the
    [plot\_hints](https://metacpan.org/pod/plot_hints) will be defined at object creation, but at
    present those settings may be overridden here and given as
    first and second arguments.

    This should be used at the end of the program (internally 
    it does an "exec").

- classify\_columns

    Given a reference to the tabular data in the form of an array of arrays,
    returns metadata for each column to be used in deciding how to plot 
    the data.

    Example usage:

        my $plot_cols = $self->classify_columns( $data );

    Classify the columns from the tabular data, returning a "fields\_metadata" hash ref.

    This is a wrapper around a provisional technique to make it easier to swap in 
    better ones later.

    At present, the metadata fields are:

         x           => $x_field  (( rename indie_x ))
         y           => $y_field
         gb_cats      => [ @gb_cats ]
         dependents_y => [ @dependents_y ]

- dumporama

    Report on state of object fields.

- fryhash

# NOTES 

## TODO

- Limited to two group by categories (in addition to the x-axis): used with colour & shape
If there's more than 2, fuse them together into a compound, use with colour
- See R Graphics Cookbook, p.205: setting up the tics and labels.

        $pc .= 'p + scale_x_date';
        $pc .= '';

- Currently this defaults to viewing images using the "display" program.
Alternately, the builder\_image\_viewer could scan through a list of 
likely viewers and pick the first that's installed.

# AUTHOR

Joseph Brenner, <doom@kzsu.stanford.edu>,
16 Nov 2016

# COPYRIGHT AND LICENSE

Copyright (C) 2016 by Joseph Brenner

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.


Changes
Revision history for Graphics-Skullplot

{{$NEXT}}

0.02  2018-11-15 18:30:57 PST
        - Initial release


scripts/skullplot.el
;;; skullplot.el --- graphical plots of sql select results

;; Copyright 2016 Joseph Brenner
;;
;; Author: doom@kzsu.stanford.edu
;; Version: $Id: skullplot.el,v 0.0 2016/07/16 23:33:00 doom Exp $
;; Keywords:
;; X-URL: not distributed yet

;; License: the same as GNU Emacs, currently GPL v.3  (see LEGAL below).

;; Description:
;; A typical database shell (e.g. psql, mysql) displays the
;; results of queries in a tabular text format I call the box format
;; or "dbox" (file extension: *.dbox).

;; The skullplot.el package exists to assist in generating graphical
;; plots of SQL query results in this "dbox" form form (by the way:
;; database purists dislike pronouncing SQL as 'sequel', an obvious
;; alternative is 'skull').  It works with a perl script skullplot.pl
;; and a perl module Data::BoxFormat, and the R statistical language package
;; ggplot2.

;; (( TODO: every piece should check that all the other pieces are installed. ))

;;  The main user function here is:
;;   skullplot-of-this-dbox

;; Installation:

;; Put this file into your load-path and the following into your ~/.emacs:
;;   (require 'skullplot)

;; TODO key assignment for skullplot-of-this-dbox in various shells and db interfaces...

;;; Code:

(provide 'skullplot)
(eval-when-compile
  (require 'cl))

;;---------
;; early function definitions

(defun skullplot-fixdir (dir &optional root)
  "Fixes the DIR.
Conditions directory paths for portability and robustness.
Some examples:
 '~/tmp'             => '/home/doom/tmp/'
 '~/tmp/../bin/test' => '/home/bin/test/'
Note: converts relative paths to absolute, using the current
default-directory setting, unless specified otherwise with the
ROOT option, this has the side-effect of converting the empty
string into the default-directory or the ROOT setting."
  (let ((return
         (substitute-in-file-name
          (convert-standard-filename
           (file-name-as-directory
            (expand-file-name dir root))))))
    return))

;;---------
;; settings and globals

(defvar progname "skullplot"
  "Name of this program (used in warnings and so on).")

(defvar skullplot-clean-up-buffers t
  "Clean-up the temporary buffers used to pass dbox data.
Defaults to t, set to nil for debugging.")
(setq skullplot-clean-up-buffers nil)

(defvar skullplot-dbox-location ""
  "Location to place the *.dbox files")
(setq skullplot-dbox-location "~/.skullplot/dbox")
(skullplot-fixdir skullplot-dbox-location)

(defvar skullplot-hr-pattern ""
  "Matches 'horizontal rule' lines used in dbox format.
This matches lines like:
 ----+------------+-----------+--------
and also the psql unicode lines:
 ────┼────────────┼───────────┼────────
" )
(setq skullplot-hr-pattern "^[-─+┼][-─+┼][-─+┼]+$")

;; TODO
;; Something like pg format might have a *single* veritcal bar
;; if you've just done a select on two fields (and actually,
;; *none* is possible, if you select just one field... but
;; you're unlikely to want to plot that, correct?)
(defvar skullplot-vbar-pattern ""
  "Match lines that have any vertical bars on them." )
(setq skullplot-vbar-pattern "^.*?[|│]")

(defvar skullplot-vbar-or-cross-pattern ""
  "Heuristic match for any dbox lines, either data, header, or hrs." )
(setq skullplot-vbar-or-cross-pattern "^.*?[|│+┼]")

;; unused
(defvar skullplot-blank-line-pattern ""
  "Match a blank line." )
(setq skullplot-blank-line-pat "^[ \t]*$")


;;---------
;; dbox plotters
;; the main user command(s)

;; wrapper to use as entry point (?)  TODO
(defun skullplot-basic-group-by ( indie-count )
  "Takes a numeric argument to distinguish dependent columns from independent.
See section 'TODO'.  The argument defaults to 1."
  (interactive "p")
  ;; (message "indie-count: %d" indie-count )
  (skullplot-of-this-dbox indie-count)
 )

(defun skullplot-of-this-dbox ( indie-count )
  "Pop a window with a graphical display of data in the current
or immediately previous dbox."
  (interactive "p")
  (let* (
         ;; emacs uses exec-path list, which has the PATH added to it
         (progfile "skullplot.pl")
         (plotter    progfile)
         (tmpfile (skullplot-create-unique-tempfile))
         (orig-buff (current-buffer))
           buffy
        )

    (setq cmd
          (cond (indie-count
                 ;; skullplot.pl --indie_count=2 input_data.dbox
                 (concat plotter " --indie_count="
                         (number-to-string indie-count)
                         " " tmpfile) )
                 (t
                  (concat plotter " " tmpfile) )
                ))

    (setq table (skullplot-get-dbox))
    (find-file tmpfile)

    (insert table)
    (save-buffer)
    (setq buffy (buffer-name))
    ;; (delete-window)  ;; ?

    (switch-to-buffer orig-buff)

    (if skullplot-clean-up-buffers
         (kill-buffer buffy))

    ;; run the cmd asynchronusly in another process (no shell):
    (start-process "skullplot" "*skullplot*" plotter tmpfile)
    (message "Running: %s" cmd);; DEBUG
    ))

(defun skullplot-of-region (beg end)
  "Pop a window with a graphical display of data in dbox in region.
You might use this if the skullplot-of-this-dbox heuristics fail,
and you want to specify what to plot manually."
  (interactive "r") ;; r -- Region: point and mark as 2 numeric args, smallest first.  Does no I/O.

  (let* (
         ;; emacs looks through exec-path list, which has the PATH added.
         (progfile "skullplot.pl")
         (slash "/") 
         (plotter      (concat progfile) )
         ;; (tmpdir       "/home/doom/tmp")  ;; Maybe "$HOME/.skullplot"?  But: create if not there
         (tmpdir   temporary-file-directory) ;; "/tmp"
         (tmpfile  (skullplot-fixdir (concat tmpdir slash "skullplot.dbox") ))
         (cmd (concat plotter " " tmpfile) )
        )
    (setq table (buffer-substring-no-properties beg end))
    (find-file tmpfile)
    ;; clear the buffer
    (mark-whole-buffer)
    (delete-region (mark) (point))
    ;;   TODO extra credit: add safety features, e.g. preserve any existing content

    (insert table)

    (save-buffer)
    ;;(write-file nil)
    ;; TODO close the buffer?  Or just the window?

    ;; run the cmd in another process (no shell):
    (message "Running: %s" cmd)

    ;; wasn't working, not sure why:
    ;; (start-process "skullplot" "*skullplot*" cmd)

    ;; works, but it blocks:
    ;; (call-process plotter tmpfile '("*skullplot*" t) t)

    ;; asynchronus
    (start-process "skullplot" "*skullplot*" plotter tmpfile)
))

;;---------
;; move to pattern

;;--------
;; hr rule

(defun skullplot-move-up-to-hr ()
  "Move upwards to first match of 'skullplot-hr-pattern' or start of buffer."
  (interactive)
  ;; Q: why not re-search-backward?  Eaiser to negate the pattern match, right?
  (move-beginning-of-line 1)
  (catch 'FLY
    (while (not (looking-at skullplot-hr-pattern ) )
      (forward-line -1) ;; move up to previous line
      (if (= (point) (point-min))
          (throw 'FLY nil))
      )))

(defun skullplot-move-down-to-hr ()
  "Move downwards to first match of 'skullplot-hr-pattern' or end of buffer.
Returns nil if not found."
  (interactive)
  ;; Q: why not re-search-forward?  A: eaiser to negate the pattern match, right?
  (move-beginning-of-line 1)
  (catch 'FLY
    (while (not (looking-at skullplot-hr-pattern ) )
      (forward-line 1)
      (move-end-of-line 1)
      (if (= (point) (point-max))
          (throw 'FLY nil))
      (move-beginning-of-line 1)
     )))

;;--------
;; vb sep lines or hrs

(defun skullplot-move-down-to-end-of-dbox ()
  "Move downwards to last line with vertical bars."
  (interactive)
  (move-beginning-of-line 1)
  (catch 'FLY
    (while ( looking-at skullplot-vbar-or-cross-pattern )
      (forward-line 1)
      (move-end-of-line 1)
      (if (= (point) (point-max))
          (throw 'FLY nil))
      (move-beginning-of-line 1)
     ))
  (forward-line -1)
  (move-end-of-line 1) )


(defun skullplot-move-up-to-start-of-dbox ()
  "Move upwards to first line with vertical bars."
  (interactive)
  (move-beginning-of-line 1)
  (catch 'FLY
    (while ( looking-at skullplot-vbar-or-cross-pattern )
      (forward-line -1) ;; move up to previous line
      (if (= (point) (point-min))
          (throw 'FLY nil))
      ))
  (forward-line 1))

;;--------
;; scraping dboxes

(defun skullplot-find-dbox ()
  "Look upwards for the previous dbox (if any)."
  (interactive)
  (let* ( (savepoint (point))
          dbox-start dbox-end  ;; start of upper line; end of lower line
          )
    (skullplot-move-up-to-hr)
    (skullplot-move-up-to-start-of-dbox)
    (setq dbox-start (point))
    (skullplot-move-down-to-end-of-dbox)
    (move-end-of-line 1)
    (setq dbox-end (point))
    ;; (message "%d %d" dbox-start dbox-end)
    (goto-char savepoint)
    (list dbox-start dbox-end)
    ))

(defun skullplot-get-dbox ()
  "Return previous dbox as string.."
  (interactive)
  (let* (( pair (skullplot-find-dbox) )
         ( beg (nth 0 pair) )
         ( end (nth 1 pair) )
         )
    (buffer-substring-no-properties beg end)))

;;--------
;; utilites

(defun skullplot-snag-line ()
  "Returns the current line as a string."
  (let ( start end line )
    (move-beginning-of-line 1)
    (setq start (point))
    (move-end-of-line 1)
    (setq end (point))
    (setq line (buffer-substring start end))
    line))

(defun skullplot-create-unique-tempfile ( &optional seed )
  "Create a uniquely named dbox file, using the SEED.
Location defaults to skullplot-dbox-location."
  (unless seed (setq seed ""))
  (let ( slashy temploc prefix templfile )
    (setq slashy (convert-standard-filename "/"))
    (setq temploc skullplot-dbox-location)
    (unless (file-exists-p temploc)
      (make-directory temploc t))
    (setq prefix
          (concat temploc slashy seed))
    (setq tempfile
          (make-temp-file prefix nil ".dbox"))
    )
  tempfile)



;; LEGAL

;; This code is licensed in the same way as GNU Emacs, which is currently
;; under the GPL v.3 (or later).

;;; skullplot.el ends here


scripts/skullplot.pl
#!/usr/bin/perl
# skullplot.pl                   doom@kzsu.stanford.edu
#                                27 Jul 2016

=head1 NAME

skullplot.pl - plot data from a manual db select

=head1 SYNOPSIS

  # default to first column on x-axis, all following columns on y-axis
  skullplot.pl input_data.dbox

  # specifying that explicitly
  skullplot.pl --dependents='x_axis_field' --independents='y_axis_field1,y_axis_field2' input_data.dbox

  # additional "dependents" fields determine color/shape of points
  skullplot.pl --dependents='x_axis_field1,category_field1' --independents='y_axis_field1,y_axis_field2' input_data.dbox

  # compact way of specifying similar case: first two columns independent, remaining columns dependent
  skullplot.pl --indie_count=2 input_data.dbox

  # don't use /tmp as working area
  skullplot.pl --working_loc='/var/scratch'  input_data.dbox

  # turn on debugging
  skullplot.pl -d  input_data.dbox

=head1 DESCRIPTION

B<skullplot.pl> is a script which use's the R ggplot2 library to
plot data input in the form of the output from a SELECT as
performed manually in a db shell, e.g.:

  +------------+---------------+-------------+
  | date       | type          | amount      |
  +------------+---------------+-------------+
  | 2010-09-01 | factory       |   146035.00 |
  | 2010-10-01 | factory       |   208816.00 |
  | 2011-01-01 | factory       |   191239.00 |
  | 2010-09-01 | marketing     |   467087.00 |
  | 2010-10-01 | marketing     |   409430.00 |
  +------------+---------------+-------------+

I call this "box format data" (file extension: *.dbox).

This script takes the name of the *.dbox file containing the data as an argument.
It has a number of options that control how it uses the data


The second argument is a comma-separated list of names of dependent variables
(the x-axis).
The third argument is a comma-separated list of the independent variables to
plot (the y-axis).

The default for dependent variables: the first column.
The default of independent variables: all of the following columns

The supported input data formats are as in the L<Table::BoxFormat> module.
At present, this is mysql and postgresql (including the unicode form).

=cut

use warnings;
use strict;
$|=1;
use Carp;
use Data::Dumper;

use File::Path      qw( mkpath );
use File::Basename  qw( fileparse basename dirname );
use File::Copy      qw( copy move );
use autodie         qw( :all mkpath copy move ); # system/exec along with open, close, etc
use Env             qw( HOME );
use Getopt::Long    qw( :config no_ignore_case bundling );
use List::Util      qw( first max maxstr min minstr reduce shuffle sum );

use utf8::all;

our $VERSION = 0.01;
my  $prog    = basename($0);

my $DEBUG   = 0;
my $working_area = "$HOME/.skullplot";   # default
my ($dependent_requested, $independent_requested, $indie_count, $image_viewer);

GetOptions ("d|debug"       => \$DEBUG,
            "v|version"     => sub{ say_version(); },
            "h|?|help"      => sub{ say_usage();   },

           "indie_count=i"  => \$indie_count,      # alt spec: indies=x1+gbcats; residue are ys
           "image_viewer=s" => \$image_viewer,     # default: ImageMagick's display (if available)
           "working_area=s" => \$working_area,

           ## Experimental, alternate interface
           "dependents=s"   => \$dependent_requested,   # the x-axis, plus any gbcats
           "independents=s" => \$independent_requested, # the y-axis
           ) or say_usage();

### $image_viewer ||= 'display';

mkpath( $working_area ) unless( -d $working_area );

# TODO dev only: remove when shipped.
use FindBin qw( $Bin );
use lib ("$Bin/../lib/",
         "$Bin/../../Data-BoxFormat/lib",
         "$Bin/../../Graphics-Skullplot/lib");

use Table::BoxFormat;
use Graphics::Skullplot;

my $dbox_file = shift;

unless( $dbox_file ) {
  die "An input data file (*.dbox) is required.";
}

if( $dependent_requested && not( $independent_requested ) ) {
  die "When using dependents option, also need independents.";
} elsif( $independent_requested && not( $dependent_requested ) ) {
  die "When using independents option, also need dependents.";
} elsif( $indie_count && $dependent_requested) {
  die "Use either indie_count or dependents/independents options, not both.";
}

if ( $dependent_requested ) {
  ($DEBUG) &&
    print STDERR "Using independents: $independent_requested and dependents: $dependent_requested\n";
} elsif( $indie_count )  {
  ($DEBUG) &&
    print STDERR "Given indie_count: $indie_count\n";
} else {
  ($DEBUG) &&
    print STDERR "Using default indie_count of 1\n";
  $indie_count = 1;
}

my $opt = { indie_count      => $indie_count,
            dependent_requested   => $dependent_requested,
            independent_requested => $independent_requested,
           };

my %gsp_args = 
  ( input_file   => $dbox_file,
    plot_hints   => $opt, );
$gsp_args{ working_area } = $working_area if $working_area;
$gsp_args{ image_viewer } = $image_viewer if $image_viewer;
my $gsp = Graphics::Skullplot->new( %gsp_args );

$gsp->show_plot_and_exit();  # does an exec 

#######
### end main, into the subs

sub say_usage {
  my $usage=<<"USEME";
  $prog -[options] [arguments]

  Options:
     -d          debug messages on
     --debug     same
     -h          help (show usage)
     -v          show version
     --version   show version

TODO add additional options

USEME
  print "$usage\n";
  exit;
}

sub say_version {
  print "Running $prog version: $VERSION\n";
  exit 1;
}


__END__

=head1 NOTES

I think I like the idea of doing it like this:

  One gb_cat:   use color
  Two gb_cats:  use shape for one with fewest values, use color for the other.
                   (( but if that value count exceeds ~6, fuse both as color ))
  Three gb_cats: use shape for the one with fewest values (( ditto ))
                 fuse the remaining items into a joint string value,
                 assign that to color.

And: this logic is getting complex enough to move to a Skullplot module
of some sort, which could open the door to a seperate package again.

What name? No reason to *presume* R/ggplot2 is there? 
That could be a plug-in style choice. 


=head1 AUTHOR

Joseph Brenner, E<lt>doom@kzsu.stanford.eduE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2016 by Joseph Brenner

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.


=cut



lib/Graphics/Skullplot.pm
package Graphics::Skullplot;
use Moo;
use MooX::Types::MooseLike::Base qw(:all);

=head1 NAME

Graphics::Skullplot - Plot the result of an SQL select (e.g. from an emacs shell window)

=head1 VERSION

Version 0.01

=cut

our $VERSION = '0.01';
my $DEBUG = 0;

=head1 SYNOPSIS

   # To use this from emacs, see scripts/skullplot.el.
   # That elisp code accesses the perl script: scripts/skullplot.pl

   # the code used by skullplot.pl
   my $plot_hints = { indie_count           => $indie_count,
                      dependent_requested   => $dependent_requested,
                      independent_requested => $independent_requested,
                    };
   my %gsp_args = 
     ( input_file   => $dbox_file,
       plot_hints   => $plot_hints, );
   $gsp_args{ working_area } = $working_area if $working_area;
   $gsp_args{ image_viewer } = $image_viewer if $image_viewer;
   my $gsp = Graphics::Skullplot->new( %gsp_args );

   $gsp->show_plot_and_exit();  # does an exec 

=head1 DESCRIPTION

Graphics::Skullplot is a module that works with the result from a database 
select in the common tabular text "data box" format. It has routines 
to generate and display plots of the data in png format.

Internally it uses the L<Table::BoxFormat> module to parse the text table,
and the L<Graphics::Skullplot::ClassifyColumns> module to determine the types of the columns.

The default image viewer is the ImageMagick "display" command.

The immediate use for this code is to act as the back-end for the included 
Emacs package scripts/skullplot.el, so that database select results 
generated in an emacs shell window can be immediately plotted.  

This elisp code calls scripts/skullplot.pl, which might be used in
other contexts.

=head1 METHODS

=over

=cut

use 5.10.0;
use Carp;
use Data::Dumper;
use File::Basename  qw( fileparse basename dirname );
# use List::Util      qw( first max maxstr min minstr reduce shuffle sum );
# use List::MoreUtils qw( any zip uniq );

use Image::Magick;

use lib "../../../Table-Classify/lib";   
use lib "../../../Data-BoxFormat/lib";  

use Table::BoxFormat;
use Graphics::Skullplot::ClassifyColumns;

=item new

Creates a new Graphics::Skullplot object.
Object attributes:

=over

=item working_area

Scratch location where intermediate files are created.
Defaults to "/tmp".

=item image_viewer

Defaults to 'display', the ImageMagick viewer
(a dependency on Image::Magick ensures it's available)

=back

=cut

# required arguments to new 
has input_file => ( is => 'ro', isa => Str,      required => 1);  # must be dbox format 
has plot_hints => ( is => 'ro', isa => HashRef,  required => 1);

has working_area => ( is => 'rw', isa => Maybe[Str], default => "/tmp" );
has image_viewer => ( is => 'rw', isa => Maybe[Str], lazy => 1, builder => "builder_image_viewer" );  

# mostly for internal use
has naming         => ( is => 'rw', isa => HashRef, lazy => 1, builder => "generate_output_filenames" ); 

=item builder methods (largely for internal use)

builder_image_viewer Currently just returns a hardcoded selection
(the ImageMagick "display" program).

=cut 

sub builder_image_viewer {
  my $self = shift;
  ($DEBUG) && print STDERR "Running _builder_image_viewer... \n";
  return "display";
}

=item generate_output_filenames

Example usage: 

  # relies on object settings: "input_file" and "working area"
  my $fn = 
    generate_filenames();
  my $basename = $fn->{ base };
  # full paths to file in $working_area
  my $tsv_file  = $fn->{ tsv };  
  my $png_file  = $fn->{ png };  

=cut 

sub generate_output_filenames {
  my $self = shift;
  my $input_file   = $self->input_file   || shift;
  my $working_area = $self->working_area || shift;
  
  my $basename = basename( $input_file ); # includes file-extension

  my ($short_base, $ext);
  if( ( $short_base = $basename ) =~ s{ \. (.*) $ }{}x ) { 
    $ext = $1;
  }

  my $tsv_name     = $short_base . '.tsv';
  my $rscript_name = $short_base . '.r';
  my $png_name     = $short_base . '.png';
  
  my $tsv_file     = "$working_area/$tsv_name";
  my $rscript_file = "$working_area/$rscript_name";
  my $png_file     = "$working_area/$png_name";

  my %filenames =
    (
     base             => $basename,
     base_sans_ext    => $short_base,
     ext              => $ext,
     tsv              => $tsv_file,
     rscript          => $rscript_file,
     png              => $png_file
     );
  # $self->naming( \%filenames );  # Tue  November 13, 2018  20:04  tango
  return \%filenames;
}


=item plot_tsv_to_png

Generate the r-code to plot the tsv file data as the png file.
Takes one argument, a hash of "field metadata".  

The file names (tsv, png, plus internal formats) come from the
"naming" object field.

Example usages:  

  $self->plot_tsv_to_png( $plot_cols ); 

=cut 

sub plot_tsv_to_png {
  my $self = shift;
  my $cd   = shift; 
  my $fn   = $self->naming         || shift;

#  my ($x_field, $y_field, $gb_cats) = @{ $cd->{ qw( x  y  gb_cats ) }}; # hash slice (mangled)

  my $x_field = $cd->{ indie_x };
  my $y_field = $cd->{ y } || $cd->{ dependents_y }->[0] ;
  my $gb_cats = $cd->{ gb_cats };

  my ($gb_cat1, $gb_cat2);
  $gb_cat1 = $gb_cats->[0] if $gb_cats->[0];
  $gb_cat2 = $gb_cats->[1] if $gb_cats->[1];

  # plot code
  my $pc = 'ggplot( skull, ' ;
  $pc .= '               aes(' ;
  $pc .= "                    x = $x_field," ;
  $pc .= "                    y = $y_field, " ;
  $pc .= "                    colour = $gb_cat1," if $gb_cat1;
  $pc .= "                    shape  = $gb_cat2 " if $gb_cat2;
  $pc .= '                          ))' ;
  $pc .= ' + geom_point( ' ;
  $pc .= "              size  = 2.5 " ;
  $pc .= '              )  ' ;

  $self->generate_png_file( $pc, $fn );
}

=item generate_png_file

Example usage:

  $self->generate_png_file( $pc, $fn );

Runs the given plot code (first argument) using the file-name metadata
(second argument, defaults to object's L<naming>), saving the 
plot as a png file ($fn->{png}).

This generates a file of R code to run with an Rscript call.
In debug mode, this generates a standalone unix script. ($DEBUG).

=cut

sub generate_png_file {
  my $self = shift;
  my $pc   = shift;
  my $fn   = shift || $self->naming; 

  my $tsv_file     = $fn->{ tsv };
  my $rscript_file = $fn->{ rscript };
  my $png_file     = $fn->{ png };

  # Generate the file of R code to run with an Rscript call
  # (in debug mode, make it a standalone unix script)
  my $r_code;
  $r_code = qq{#!/usr/bin/Rscript} . "\n" if $DEBUG;

  $r_code .=<<"__END_R_CODE";
library(ggplot2)
skull <- read.delim("$tsv_file", header=TRUE)
png("$png_file") # send plot output to png
$pc
graphics.off()   # doesn't chatter like dev.off
__END_R_CODE

  print $r_code, "\n" if $DEBUG;

  open my $out_fh, '>', $rscript_file;
  print { $out_fh } $r_code;
  close( $out_fh );

  # in case you want to run the rscript standalone
  chmod 0755, $rscript_file;

   # chdir( "$HOME/tmp" ) or die "$!";

  my $erroff = '2>/dev/null';
  $erroff = '' if $DEBUG;

  my $cmd = "Rscript $rscript_file $erroff";

  print STDERR "cmd:\n$cmd\n" if $DEBUG;
  system( $cmd );
}




=item display_png_and_exit

Open the given png file in an image viewer
Defaults to "png" field in object's "naming".

This internally does an exec: it should be
the last thing called.

The image viewer can be set as the second, optional field.
The default image viewer is ImageMagick's "display".

Example usage:

  my $naming = $self->naming;
  my $png_file = $naming->{ png };
  $self->display_png_and_exit( $png_file );

=cut

sub display_png_and_exit {
  my $self        = shift;
  my $fn          = $self->naming; 
  my $png_file    = shift || $fn->{ png };
  my $basename = $fn->{ base } || '';

  my $image_viewer = $self->image_viewer || shift;

  my $erroff = ' 2>/dev/null';
  $erroff = '' if $DEBUG;

  my $vcmd;
  if ($image_viewer eq 'display')  {
    my $title = "skullplot";
    # $title .=  ": $basename" if $basename;
    $vcmd = qq{ display -title $title  $png_file $erroff };
  } else {
    $vcmd = qq{ $image_viewer $png_file $erroff };
  }
  exec( $vcmd );
}



=item show_plot_and_exit

The method called by the skullplot.pl script to actually
plot the data from a "data box format" file, using the 
plot_hints.

It's expected that the dbox file (L<input_file>) and the
L<plot_hints> will be defined at object creation, but at
present those settings may be overridden here and given as
first and second arguments.

This should be used at the end of the program (internally 
it does an "exec").

=cut

sub show_plot_and_exit {
  my $self = shift;
  if ($DEBUG) { say "Running show_plot_and_exit: "; $self->dumporama() };

  my $dbox_file = $self->input_file || shift;

  my $naming = $self->naming;

  my $dbox_name = $naming->{ base };
  my $tsv_file  = $naming->{ tsv };
  ($DEBUG) && print "input dbox name: $dbox_name\nintermediate tsv_file: $tsv_file\n";

  # the input from the dbox file output directly to a tsv file 
  my $dbx = Table::BoxFormat->new( input_file  => $dbox_file ); 
  my $data = $dbx->output_to_tsv( $tsv_file ); # also returns a ref to an array of arrays

  my $plot_cols = $self->classify_columns( $data );

  $self->plot_tsv_to_png( $plot_cols ); # Note: uses naming from object

  if ($DEBUG) { say "About to display png: "; $self->dumporama() };
  $self->display_png_and_exit(); 
}



=item classify_columns

Given a reference to the tabular data in the form of an array of arrays,
returns metadata for each column to be used in deciding how to plot 
the data.

Example usage:

  my $plot_cols = $self->classify_columns( $data );


Classify the columns from the tabular data, returning a "fields_metadata" hash ref.

This is a wrapper around a provisional technique to make it easier to swap in 
better ones later.

At present, the metadata fields are:

     x           => $x_field  (( rename indie_x ))
     y           => $y_field
     gb_cats      => [ @gb_cats ]
     dependents_y => [ @dependents_y ]

=cut

sub classify_columns {
  my $self = shift;
  my $data = shift;  

  my $opt         = $self->plot_hints || shift;
  # my $indie_count = $opt->{ indie_count };  

  # use the tsv data to analysis the column types, 
  # determine what to try to plot
  my $dc = Graphics::Skullplot::ClassifyColumns->new( data => $data );  
  my $plot_cols = 
    $dc->classify_columns_simple( $opt );

  return $plot_cols;
}





=item dumporama

Report on state of object fields.

=cut

sub dumporama {
  my $self = shift;
  say "Graphics::Skullplot self: ", Dumper( $self );
  printf "input_file: %s\n"     , $self->input_file;
  # printf "plot_hints: %s\n"     , $self->fryhash( $self->plot_hints );
  printf "plot_hints: %s\n"     , Dumper( $self->plot_hints );
  printf "working_area: %s\n"   , $self->working_area;
  printf "image_viewer: %s\n"   , $self->image_viewer;
  # printf "naming: %s\n"         , $self->fryhash( $self->naming );
  printf "naming: %s\n"         , Dumper( $self->naming );
}



=item fryhash

=cut

sub fryhash {
  my $self = shift;
  my $href = shift;

  my @keys = keys %{ $href };
  my $rep;
  $rep = "\n" if @keys;
  foreach my $k (@keys) {
    my $val = $href->{ $k } || '';
    if( ref $val eq 'HASH') {
      $rep .= $self->fryhash( $val );
    } elsif( ref $val eq 'ARRAY') {
      $rep .= join " ", @{ $val };
    } else {
      $rep .= sprintf "         %25s: %s\n", $k, $val;
    }
  }
  return $rep;
}



=back

=head1 NOTES 

=head2 TODO

=over 

=item * 

Limited to two group by categories (in addition to the x-axis): used with colour & shape
If there's more than 2, fuse them together into a compound, use with colour

=item * 

See R Graphics Cookbook, p.205: setting up the tics and labels.

    $pc .= 'p + scale_x_date';
    $pc .= '';

=item * 

Currently this defaults to viewing images using the "display" program.
Alternately, the builder_image_viewer could scan through a list of 
likely viewers and pick the first that's installed.

=back


=head1 AUTHOR

Joseph Brenner, E<lt>doom@kzsu.stanford.eduE<gt>,
16 Nov 2016

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2016 by Joseph Brenner

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut

1;
